sdk/include/reactos/imm32_undoc.h

```diff
@@ -2,11 +2,83 @@
  * PROJECT:     ReactOS Kernel
  * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
  * PURPOSE:     Private header for imm32.dll
- * COPYRIGHT:   Copyright 2021 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
+ * COPYRIGHT:   Copyright 2021-2023 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
  */
 
 #pragma once
 
+typedef struct tagTRANSMSG
+{
+    UINT message;
+    WPARAM wParam;
+    LPARAM lParam;
+} TRANSMSG, *PTRANSMSG, *LPTRANSMSG;
+
+typedef struct tagTRANSMSGLIST
+{
+    UINT     uMsgCount;
+    TRANSMSG TransMsg[ANYSIZE_ARRAY];
+} TRANSMSGLIST, *PTRANSMSGLIST, *LPTRANSMSGLIST;
+
+typedef struct tagIMEINFOEX
+{
+    HKL hkl;
+    IMEINFO ImeInfo;
+    WCHAR wszUIClass[16];
+    ULONG fdwInitConvMode;
+    INT fInitOpen;
+    INT fLoadFlag;
+    DWORD dwProdVersion;
+    DWORD dwImeWinVersion;
+    WCHAR wszImeDescription[50];
+    WCHAR wszImeFile[80];
+    struct
+    {
+        INT fSysWow64Only:1;
+        INT fCUASLayer:1;
+    };
+} IMEINFOEX, *PIMEINFOEX;
+
+typedef enum IMEINFOEXCLASS
+{
+    ImeInfoExKeyboardLayout,
+    ImeInfoExKeyboardLayoutTFS,
+    ImeInfoExImeWindow,
+    ImeInfoExImeFileName
+} IMEINFOEXCLASS;
+
+/* unconfirmed */
+typedef struct tagCLIENTIMC
+{
+    HANDLE hInputContext;   /* LocalAlloc'ed LHND */
+    LONG cLockObj;
+    DWORD dwFlags;
+    DWORD dwCompatFlags;
+    RTL_CRITICAL_SECTION cs;
+    UINT uCodePage;
+    HKL hKL;
+    BOOL bUnknown4;
+} CLIENTIMC, *PCLIENTIMC;
+
+#define DEFINE_IME_ENTRY(type, name, params, extended) typedef type (WINAPI *FN_##name) params;
+#include "../../../win32ss/include/imetable.h"
+#undef DEFINE_IME_ENTRY
+
+typedef struct IMEDPI
+{
+    struct IMEDPI *pNext;
+    HINSTANCE      hInst;
+    HKL            hKL;
+    IMEINFO        ImeInfo;
+    UINT           uCodePage;
+    WCHAR          szUIClass[16];
+    DWORD          cLockObj;
+    DWORD          dwFlags;
+#define DEFINE_IME_ENTRY(type, name, params, extended) FN_##name name;
+#include "../../../win32ss/include/imetable.h"
```

Comment:
This would give yet another reason why the whole imm32 and related IMM modules should be moved into win32ss/ .

Reply:
`imetable.h` and `immtable.h` are moved.